tools/doccheck: extend script to also check for Doxygen groups defined multiple times

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

While reworking #8939, I noticed several nucleo boards were having their Doxygen groups (e.g. `boards_nucleo-xxx`) defined in `doc.txt` and `periph_conf.h`.

After looking into it a bit more, it appears that some other Doxygen groups, in different parts of the code, have the same problem.

The idea of this PR is to extend the doccheck script to also check groups that are defined multiple times and to return the number of times and files where the groups are defined.

For the moment, the script doesn't exit with a return code != 0 in case of problem but simply prints the problematic groups with a warning. So it can be merged without bad consequences on static tests.

Previous checks still work (according to my local tests).

I'll open another PR with all reported doxygen group fixed and a change to make the script return an error if one or several groups are defined multiple times.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

- Run `make static-test` and check the output
- Verify the listed groups are indeed problematic
- CI should not fail

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

~The PR containing all the fixes will arrive soon.~ All problems found are fixed by #11814 

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

cladmi commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +76,47 @@ then
+
+# Check for groups defined multiple times
+MULTIPLE_DEFINED_GROUPS=$( \
+    for group in $(echo ${DEFINED_GROUPS_UNIQUE} | tr ' ' '\n'); \
```
Would a `echo ${DEFINED_GROUPS} | sort | uniq -c` give a good enough output ?

aabadie replied:
I tried it and could make it work.

cladmi commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +76,47 @@ then
+# Expected output lines patterns:
+# <filepath>: * <group name>
+# <filepath>:<group name>
+ALL_RAW_DEFGROUP=$(git grep '@defgroup' -- '*.h' '*.c' '*.txt' | exclude_filter | \
```
This is calling `git grep @defgroup` a second time.

aabadie replied:
I reworked the script so that git grep is only called once (for each) to find ingroup and defgroup uses.

cladmi commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +76,47 @@ then
+        _group_name=$(echo ${group} | awk -F: '{ printf ("%s\n",$2) }')
+        _group_count=$(echo ${group} | awk -F: '{ printf ("%s\n",$1) }')
+        echo -e "\n${CWARN}${_group_name}${CRESET} defined ${CWARN}${_group_count}${CRESET} times in:"; \
+        echo "${ALL_RAW_DEFGROUP}" | grep "\<${_group_name}\>$" | sort -u | \
```
Oh there was a `\` at the end but the line is not indented.

aabadie commented:
Thanks for having a first look @cladmi.

> No ideas about how doxygen groups should be handled.

I tend to think that there should only one call to defgroup for a given group, even if Doxygen doesn't raise a warning when there are duplicate defgroup.

Your previous comments should be addressed now. I reworked a bit the initial version of this PR following them.

miri64 commented:
Please rebase and squash so I can have another look with #11814.

aabadie commented:
Done

miri64 commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +75,40 @@ then
+# - remove empty spaces from the beginning of the line
+# - replace spaces with ':' for easily looping on the result when printing
+# - lines starting with '1:' are excluded
+MULTIPLE_DEFINED_GROUPS=$(echo "${DEFINED_GROUPS}" | uniq -c | \
```
Do they have to be counted? If not, you can simplify this operation to:

```suggestion
MULTIPLE_DEFINED_GROUPS=$(echo "${DEFINED_GROUPS}" | uniq -d)
```

From [specificaiton]:

```
-d
    Suppress the writing of lines that are not repeated in the input.
```

miri64 commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +75,40 @@ then
+    do \
+        _group_name=$(echo ${group} | awk -F: '{ printf ("%s\n",$2) }'); \
+        _group_count=$(echo ${group} | awk -F: '{ printf ("%s\n",$1) }'); \
+        echo -e "\n${CWARN}${_group_name}${CRESET} defined ${CWARN}${_group_count}${CRESET} times in:"; \
```
This is the only instance count is used and it could be replaced with "multiple", as you list them anyway below, so I think the simplicfication I proposed above can be applied.

miri64 commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +75,40 @@ then
+        _group_name=$(echo ${group} | awk -F: '{ printf ("%s\n",$2) }'); \
+        _group_count=$(echo ${group} | awk -F: '{ printf ("%s\n",$1) }'); \
+        echo -e "\n${CWARN}${_group_name}${CRESET} defined ${CWARN}${_group_count}${CRESET} times in:"; \
+        echo "${ALL_DEFGROUP_CLEANED}" | grep "\<${_group_name}\>$" | sort -u | \
```
I think it is more efficient to create a group specific `DEFGROUP_CLEANED` directly here. You only use `ALL_DEFGROUP_CLEANED` here but it contains also groups that are not duplicates, so all the operations [here](https://github.com/RIOT-OS/RIOT/pull/11813/files#diff-68c36a96890fc4b2b308e44462797902R93-R94) are always executed, even if `ALL_DEFGROUP_CLEANED` is not used.

miri64 replied:
So something like

```suggestion
        echo "${ALL_RAW_DEFGROUP}" | grep "\<${_group_name}\>$" | \
            awk -F@ '{ split($2, end, " "); printf("\t%s\n",$1) }' | sort -u
```

(and remove `ALL_DEFGROUP_CLEANED` of course)

miri64 commented on dist/tools/doccheck/check.sh:
```diff
@@ -74,3 +75,40 @@ then
+    COUNT=$(echo "${MULTIPLE_DEFINED_GROUPS}" | wc -l)
+    echo -ne "${CERROR}ERROR${CRESET} "
+    echo -e "There are ${CWARN}${COUNT}${CRESET} Doxygen groups defined multiple times:"
+    echo "${MULTIPLE_GROUPS_PRINT}"
```
Why not just put the loop here instead of assigning its result to a variable you print anyway?

aabadie commented:
> Do they have to be counted?

If you think they are not needed, then i'm fine with removing them. It will simplify quite a few this PR (it was a nightmare to make it work correctly with the counts ;) ).

miri64 commented:
> > Do they have to be counted?
> 
> If you think it's not needed, then i'm fine with removing the counts. It will simplify quite a few this PR (it was a nightmare to make it work correctly with the counts ;) ).

I always prefer exact numbers over no numbers, however in this case I don't think they justify the code complexity that is introduced here ;-).

aabadie commented:
@miri64. I reworked the script like you suggested in a fixup commit. I added a second commit with the same print improvement for undefined groups (only compute the list of groups and files if there are undefined groups found).

miri64 commented on dist/tools/doccheck/check.sh:
```diff
@@ -71,6 +64,30 @@ then
+then
+    COUNT=$(echo "${MULTIPLE_DEFINED_GROUPS}" | wc -l)
+    echo -ne "${CERROR}ERROR${CRESET} "
+    echo -e "There are ${CWARN}${COUNT}${CRESET} Doxygen groups defined multiple times:"
```
:-)

miri64 commented:
Will look into it in detail and test tomorrow.

aabadie commented:
@miri64, merge ?

miri64 commented:
Yepp, sorry, lost track.
